astropy/io/fits/tests/test_table.py

```diff
@@ -3035,6 +3035,31 @@ def test_tolist_from_file(self):
             assert hdu.data.tolist() == [[[45, 56], [11, 3]], [[11, 12, 13], [12, 4]]]
             assert hdu.data['var'].tolist() == [[45, 56], [11, 12, 13]]
 
+    def test_heapsize_P_limit(self):
+        """
+        Regression test for https://github.com/astropy/astropy/issues/10812
+
+        Check if the error is raised when the heap size is bigger than what can be
+        indexed with a 32 bit signed int.
+        """
+
+        # a matrix with variable length array elements is created
+        nrows = 30000
+        matrix = np.zeros(nrows, dtype=np.object_)
+        for i in range(0, nrows):
+            matrix[i] = np.arange(0., float(i+1))
```

Comment:
This seems like a very expensive test, actually filling about a GB of memory just to check an error message gets created. Would it be possibly to check this somewhat more easily? If this really is the easiest way, I'd suggest to use 
```
matrix[i] = np.broadcast_to(np.zeros(()), (i+1,))
```
instead, which will produce an array that looks like it has `i+1` elements, but does not take up the memory. Ideally rewritten with far fewer rows to speed up matters, as in
```
matrix = np.empty(2, dtype=np.object_)
matrix[0] = np.broadcast_to(np.zeros(()), (30000, 30000))
matrix[1]] = np.arange(10.)
```
(didn't test whether this causes a problem, but assuming it is just a question of making a single offset too big)

Reply:
I will check... Will this fix the 32bit failure too?
Thanks for your suggestion.

Reply:
Hmm, that will need more of a `io.fits` expert than I am. I was just trying to make the test less expensive! Let's ping @saimn...

Reply:
There is another test using a lot of memory in `io.ascii`, and for this test we use the `@pytest.mark.hugemem` marker. Those tests are not run on the CI because of the big memory requirement. Not use what is the limit on the CI providers but the idea was that we don't need to burn CI time for those tests and can run them manually, at least before doing a release. I guess we could do the same here ?

Reply:
To simplify things I think we could have a single row with an zeros array of 2Gb.